modules/mauth-authenticator-http4s/src/main/scala/com/mdsol/mauth/http4s/RequestAuthenticator.scala

```diff
@@ -0,0 +1,151 @@
+    * The validation process consists of recreating the mAuth hashed signature from the request data
+    * and comparing it to the decrypted hash signature from the mAuth header.
+    *
+    * @param mAuthRequest Data from the incoming HTTP request necessary to perform the validation.
+    * @return True or false indicating if the request is valid or not with respect to mAuth.
+    */
+  override def authenticate(mAuthRequest: MAuthRequest)(implicit requestValidationTimeout: Duration): F[Boolean] = {
+    validateTime(mAuthRequest.getRequestTime, epochTimeProvider)(requestValidationTimeout)
+      .flatMap {
+        case false =>
+          val message = s"MAuth request validation failed because of timeout $requestValidationTimeout"
+          Logger[F].error(message) *> ApplicativeThrow[F].raiseError(new MAuthValidationException(message))
+        case true =>
+          validateMauthVersion(mAuthRequest, v2OnlyAuthenticate).flatMap {
+            case false =>
+              val message = "The service requires mAuth v2 authentication headers."
+              Logger[F].error(message) *> ApplicativeThrow[F].raiseError(new MAuthValidationException(message))
+            case _ => getPublicKey(mAuthRequest)
+          }
+      }
+  }
+
+  private def getPublicKey(mAuthRequest: MAuthRequest): F[Boolean] = {
+    publicKeyProvider.getPublicKey(mAuthRequest.getAppUUID).flatMap {
+      case None =>
+        Logger[F].error("Public Key couldn't be retrieved") *> false.pure[F]
+      case Some(clientPublicKey) =>
+        // Decrypt the signature with public key from requesting application.
+        mAuthRequest.getMauthVersion match {
+          case MAuthVersion.MWS =>
+            Logger[F].warn("MAuth v1 client was used to authenticate this request which is deprecated") *>
+              validateSignatureV1[F](mAuthRequest, clientPublicKey)
+          case MAuthVersion.MWSV2 if isV2OnlyAuthenticate => validateSignatureV2(mAuthRequest, clientPublicKey)
+          case MAuthVersion.MWSV2 =>
+            validateSignatureV2(mAuthRequest, clientPublicKey).flatMap {
+              case true  => true.pure[F]
+              case false => fallbackValidateSignatureV1[F](mAuthRequest, clientPublicKey)
+            }
+        }
+    }
+  }
+
+}
+
+object RequestAuthenticator {
+  def apply[F[_]: Async: Logger](publicKeyProvider: ClientPublicKeyProvider[F], epochTimeProvider: EpochTimeProvider) =
+    new RequestAuthenticator(publicKeyProvider, epochTimeProvider, v2OnlyAuthenticate = false)
+
+  // Check epoch time is not older than specified interval.
+  private def validateTime[F[_]: Async](requestTime: Long, epochTimeProvider: EpochTimeProvider)(requestValidationTimeout: Duration): F[Boolean] =
+    ((epochTimeProvider.inSeconds - requestTime) < requestValidationTimeout.toSeconds).pure[F]
+
+  // Check V2 header if only V2 is required
+  private def validateMauthVersion[F[_]: Async](mAuthRequest: MAuthRequest, v2OnlyAuthenticate: Boolean): F[Boolean] =
```

Comment:
```suggestion
  private def validateMauthVersion[F[_]: Applicative](mAuthRequest: MAuthRequest, v2OnlyAuthenticate: Boolean): F[Boolean] =
```

same with regards to non-private